Lab on Git Bisect

RocketPuppy opened the pull request:
cc @joealba since you asked for it

RocketPuppy commented on src/labs.txt:
```diff
@@ -2745,6 +2745,200 @@ Output:
+=bisectstart1
+git bisect old <hash>
+=bisectstart2
+git bisect new HEAD
```
New/old fits the "when did this change happen" model better than good/bad. I actually didn't know about it until I was looking at the manual for bisect while doing this lab. Given the small scale of this project there aren't really any bugs and I didn't want to clutter an already complicated lab by adding an artificial bug so I went with the slightly less contrived approach of finding when a change happened.

cpjolicoeur replied:
I think that good/bad is the more "typical" terminology people will be using though (and do use day to day).  As you said as well, you didnt even know about new/old until researching this lab.

Bisect is (most?) often used to "hunt down where a bug was introduced", and as such framing the lab in the scope of what the normal/average usage of the command makes a bit more sense to me.

I do agree that it could seem weird to introduce an artificial bug into the lab, but I'd have to think about if I think that is a more or less contrived example to use.  I guess it's probably different sides of the same coin as we have to craft some sort of example use-case to allow the lab to exist.

RocketPuppy replied:
So my biggest concern with crafting a bug is for the bisect to demonstrate its value the bug would have to be somewhere relatively deep in the history. Starting the lab by saying, "Write this bug and commit it" would just leave bisect to point at the commit they just created and I felt like that wouldn't demonstrate the point very well. Going with `old/new` instead of `good/bad` felt like it preserved the value of the bisect better to me.

shoe replied:
Let's stick with "good"/"bad".  People can learn obscure options from the man pages.

RocketPuppy commented on src/labs.txt:
```diff
@@ -2745,6 +2745,200 @@ Output:
+
+h2. The Test Command
+
+p. The first thing we need is a test command. @git bisect@ works by checking
```
You can give bisect a script to automate it and it just uses exit 0 or exit non-0 for bad/good or old/new respectively.

cpjolicoeur replied:
I follow that, my understanding/assumption is that the more common usage pattern is for people to manually declare good/bad and not write a script.  I know I personally have only ever written a script runner probably once or twice ever and I've never taught it that way as the "default" to people

joealba replied:
I agree with Craig here.  I would still like to keep the script content, as it is awesome...  it's just better as a "you can also do this kickass thing to automate the manual thing we just showed you" add-on topic.

RocketPuppy replied:
Interesting. I write a script for almost all of my bisects. I will point out (and I understand if it's hard to see because there's no preview) that the lab does start by running the script manually and doing `git bisect old/new`. If you think it's better to leave the script off entirely until the automation step I can do that too, though I do like the idea that it encourages getting the computer to do the work for you.

shoe replied:
I'd mention the script and `bisect run` option as a detail, but the more fundamental technique is just without a script. It's often difficult to write a script for what you're looking for.  "Good" vs. "Bad" might be a visual regression, for example.

RocketPuppy commented:
@cpjolicoeur rebased, updated. This is open against my fork and not this repo. The branch is fine, but you won't get a preview for it unless I reopen.

RocketPuppy commented:
@cpjolicoeur @joealba @shoe Updated. Changed to good/bad and I think I have it worded so it doesn't come across as opposite of what is desired. Reworked the first bisect section to happen manually and moved the script to the automation section. I opened a draft PR to get a preview flow working https://git.withmojo.com/dw/bisect-lab/lab_43.html

RocketPuppy commented:
@joealba should be framed more like a bug now.

joealba commented on src/labs.txt:
```diff
@@ -2809,6 +2809,216 @@ Output:
+
+p. Bugs happen, and it can be necessary to find the cause of a bug without
+knowing what code specifically is involved. Git allows you to efficiently
+search for a commit as long as you can describe your search criteria in terms
```
Even without an executable test, it's still more efficient / helpful to use bisect to keep track of which commits were good and which commits were bad within a range.

RocketPuppy replied:
I'm not sure what you're getting at here.

joealba replied:
"Git allows you to efficiently search for a commit as long as you can describe your search criteria in terms of a test that can be executed."

"A test that can be executed" implies to me that you need some kind of runnable command to perform the test.  This sentence _could_ cover the "I'm manually refreshing my browser to see if the behavior is better now" scenario too, but I think we can reword to cover it more clearly.

RocketPuppy replied:
What about: "Git allows you to efficiently search for a commit by taking some action and telling git if the commit is closer or farther than the one you are looking for."

joealba replied:
"closer or farther" isn't really accurate -- you never really get further away from the target commit as you proceed.  It's more "before or after the change you are looking for", but that's a little awkward as an intro too.  It's hard to describe abstractly.  I feel like it is easier to clearly describe bisect's behavior specifically in the context of a bug search.

```
Bugs happen, and it can be necessary to find the cause of a bug without
knowing what code specifically is involved. Git allows you to efficiently
search a series of commits to find the commit that introduced the bug.

Git bisect asks you to declare a known good commit and a known bad commit.  Then git bisect will pick a commit in the middle of that set and ask you test that commit.  You'll test your project at that point either manually or via an automated test, and then you'll mark the commit as good or bad.  Git bisect will continue to pick the midpoint between the latest good and bad commits until you land on the exact commit that introduced the bug.
```

RocketPuppy replied:
@joealba Updated. Preview should be up to date shortly.
